MM-42819 mmctl: informative errors on permanent deletions

willypuzzle opened the pull request:
#### Summary
A new constant for the configuration documentation URL has been introduced. This URL is now included in the long description of the delete commands for channels, teams, and users. The user is informed that certain service settings must be enabled to use these commands, and directed to the documentation for more information.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/19843

#### Release Note
```release-note
NONE
```

mattermost-build commented:
Hello @willypuzzle,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

hanzei commented on server/cmd/mmctl/commands/constants/urls.go:
```diff
@@ -0,0 +1,5 @@
+package constants
```
`constants` is not a good package names (https://go.dev/blog/package-names#bad-package-names). Let's just move the `const` into the `commands` package.

hanzei commented on server/cmd/mmctl/commands/user.go:
```diff
@@ -142,7 +143,8 @@ var DeleteUsersCmd = &cobra.Command{
 	Short: "Delete users",
 	Long: `Permanently delete some users.
-Permanently deletes one or multiple users along with all related information including posts from the database.`,
+Permanently deletes one or multiple users along with all related information including posts from the database.
+In order to use this command ServiceSettings.EnableAPIUserDeletion must be set to true. See ` + constants.CONFIG_DOCUMENTATION_URL + ` for more information.`,
```
I would prefer if the command checks the config and outputs an actionable error instead of having the static documentation in the help text.

Does that sound reasonable to you?

willypuzzle replied:
ok!

willypuzzle replied:
@hanzei , the problem I'm encountering is that I should load app with the right config in order to make a conditional output. How can I achieve this?

hanzei reviewed (changes requested):
Thanks for the PR!

hanzei commented on server/cmd/mmctl/commands/channel.go:
```diff
@@ -612,6 +613,15 @@ func getPrivateChannels(c client.Client, teamID string) ([]*model.Channel, error
 }
 
 func deleteChannelsCmdF(c client.Client, cmd *cobra.Command, args []string) error {
+	app, err := commands.InitDBCommandContextCobra(cmd)
+	if err != nil {
+		printer.PrintT("App initialization failed: %s", err.Error())
+		return err
+	}
+	deleteEnabled := app.Config().ServiceSettings.EnableAPIChannelDeletion
```
That only works if `mmctl` is run on the same server as MM. Let' use `c.GetConfig` instead.

hanzei commented on server/cmd/mmctl/commands/urls.go:
```diff
@@ -0,0 +1,5 @@
+package commands
+
+const (
+	CONFIG_DOCUMENTATION_URL = "https://docs.mattermost.com/configure/environment-configuration-settings.html"
```
Please use standard CamelCase instead

hanzei reviewed (changes requested):
Much better!

hanzei commented on server/cmd/mmctl/commands/utils/main.go:
```diff
@@ -0,0 +1,22 @@
+
+func GetConfig(c client.Client, context context.Context) (*model.Config, error) {
+	config, response, err := c.GetConfig(context)
+	if err != nil || response.StatusCode < 200 || response.StatusCode >= 300 {
+		if err != nil {
+			return nil, err
+		}
+		return nil, errors.New(fmt.Sprintf("Wrong response client: code: %d", response.StatusCode))
+	}
```
`GetConfig` already checks for errors (https://github.com/mattermost/mattermost/blob/79e1c182ce435e8b946535a4344de260a31f504e/server/public/model/client4.go#L763-L798)

Do we need the custom code?

willypuzzle replied:
So, we're sure.

hanzei replied:
I think we can completely remove the `utils.GetConfig` function. `client4.GetConfig` does everything we need.

willypuzzle replied:
I did so

hanzei reviewed (commented):
CI is red. I'm unsure if that is related to the PR. merged `master` and restarted it.

hanzei commented on server/cmd/mmctl/commands/urls.go:
```diff
@@ -0,0 +1,8 @@
+package commands
+
+const (
+	ConfigDocumentationUrl = "https://docs.mattermost.com/configure/environment-configuration-settings.html"
```
@cwarnermm is there a PL we can use?

cwarnermm replied:
Requested PL link from marketing:
- route: ``https://mattermost/com/pl/environment-configuration-settings``
- destination: ``https://docs.mattermost.com/configure/environment-configuration-settings.html``

willypuzzle replied:
There is any news about it?

cwarnermm replied:
Thanks for following up, @willypuzzle! Yes, this PL is now set up and available for use in-product.

willypuzzle replied:
Oh, yes! But I don't know what a PL is and how to use it. Could you give me some explanation?

willypuzzle replied:
please @hanzei help me :)

hanzei replied:
@willypuzzle Sorry, I missed your ping. You can just use https://mattermost.com/pl/environment-configuration-settings in your code directly.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hanzei reviewed (commented):
Heads up that more tests are failing: https://github.com/mattermost/mattermost/actions/runs/15117452254/job/42530869163?pr=30230

willypuzzle commented:
@hanzei I'm facing an issue. According to e2e tests the command should ignore the configurazion if the command has a LocalClient. But I haven't found a way to check if the client is local at command time (in short when I call c.GetConfig()).
What can we do?

willypuzzle commented:
> Maybe I'm missing something here. Why should the command ignore the config if a LocalClient is used?

@hanzei I think because if one try from a remote machine then the flag should be true but if you try from the machine itself then the flag could be bypassed. Anyway, this is the behavior traced in the e2e tests, for instance https://github.com/mattermost/mattermost/blob/master/server/cmd/mmctl/commands/user_e2e_test.go#L821

hanzei commented:
Why should running mmctl on the same machine as MM cause the check to be ignored? I don't think that is how the server works.

hanzei commented:
Can you please take a look at the CI errors? Thanks!

hanzei commented:
Maybe I'm missing something here. The e2e tests should already report an error as the server checks if the config setting is enabled. It's just the error message that changes, right?

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hanzei commented:
> According to e2e tests the command should ignore the configurazion if the command has a LocalClient.

That is a great observation! I hate to say that, but I feel like the whole approach to PR is flawed. The server knows much better what the client is and what it is not allowed to do and can return the correct error message.

@willypuzzle @isacikgoz What you do think about changing the error message on the server based on the fact that a user is an admin?
Instead of `Unable to delete channel 'off-topic' error: : The permanent channel deletion function is disabled. Contact your system administrator。,` it would return `Unable to delete channel 'off-topic' error:  ServiceSettings.EnableAPIChannelDeletion must be set to true to use this command. See https://mattermost.com/pl/environment-configuration-settings for more information` for admins.

isacikgoz commented:
> we can close this PR and open another?

I think you can start over in this PR, nothing wrong with that. And for the messaging, it would help to change according to the client.

willypuzzle commented:
> Golangci is unhappy: https://github.com/mattermost/mattermost/actions/runs/15679433070/job/44168709137?pr=30230

@hanzei I modified the PR, unmasked the variable and corrected failed tests

willypuzzle commented:
@hanzei what is it supposed to do in these cases?

hanzei commented on server/channels/api4/team.go:
```diff
@@ -477,7 +477,12 @@ func deleteTeam(c *Context, w http.ResponseWriter, r *http.Request) {
-			err = model.NewAppError("deleteTeam", "api.user.delete_team.not_enabled.app_error", nil, "teamId="+c.Params.TeamId, http.StatusUnauthorized)
+			key := "api.user.delete_team.not_enabled.app_error"
+			user, usrErr := c.App.GetUser(c.AppContext.Session().UserId)
+			if usrErr == nil && user != nil && model.IsInRole(user.Roles, model.SystemAdminRoleId) {
```
We should be using `IsSystemAdmin()` here

hanzei commented on server/channels/api4/team.go:
```diff
@@ -477,7 +477,12 @@ func deleteTeam(c *Context, w http.ResponseWriter, r *http.Request) {
+			key := "api.user.delete_team.not_enabled.app_error"
+			user, usrErr := c.App.GetUser(c.AppContext.Session().UserId)
+			if usrErr == nil && user != nil && model.IsInRole(user.Roles, model.SystemAdminRoleId) {
+				key = "api.user.delete_team.not_enabled.for_admin.app_error"
```
Let's add a comment explaining the difference between the two error messages.

hanzei reviewed (commented):
Much better :rocket:

willypuzzle commented:
> @willypuzzle Could you not force push to your branch? Thanks make reviewing PRs much harder. Please check out https://mattermost.com/blog/submitting-great-prs/ for more information.

sorry, I didn't know it was annoying. I use it in other projects because rebasing to master was requested. Anyway, thanks for the information, I will avoid using it here on mattermost.

hanzei commented:
Thanks you!

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hanzei commented:
@mgdelacroix Gentle reminder to review the PR :wink:

mgdelacroix commented on server/cmd/mmctl/commands/team_e2e_test.go:
```diff
@@ -125,6 +125,51 @@ func (s *MmctlE2ETestSuite) TestDeleteTeamsCmdF() {
+		config.ServiceSettings.EnableAPITeamDeletion = &enableConfig
+		_, _, _ = c.UpdateConfig(context.TODO(), config)
+
+		// Deletion should succeed for both local and SystemAdmin client now
```
This is just testing the local client behavior, right? Deletion should fail for sysadmins when the config is disabled as the next test proves if I understood correctly

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hanzei commented on server/channels/api4/channel.go:
```diff
@@ -1366,7 +1366,12 @@ func deleteChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+			key := "api.user.delete_channel.not_enabled.app_error"
+			user, usrErr := c.App.GetUser(c.AppContext.Session().UserId)
+			if usrErr == nil && user != nil && user.IsSystemAdmin() {
+				key = "api.user.delete_channel.not_enabled.for_admin.app_error" // More verbose error message for system admins
```
Arg, the manual assignment of the translation key is causing issues with our tooling. Can you just use `model.NewAppError` to create the separate errors? Thanks!

willypuzzle replied:
ok

hanzei commented:
@willypuzzle Please run `make i18n-extract` in the server repo to update the i18n files. Thanks!

hanzei reviewed (approved):
Awesome 🚀

lindy65 reviewed (approved):
Thanks @willypuzzle, looks good 👍

hanzei commented:
/check-cla

hanzei commented:
Thanks @willypuzzle :tada:
